docs/modules/ROOT/pages/Java_Case_Study.adoc

```diff
@@ -0,0 +1,430 @@
+  def moduleDeps = Seq(common, buffer, transport, `transport-native-unix-common`, `transport-classes-kqueue`)
+  def testModuleDeps = Seq(testsuite, `transport-native-unix-common-tests`)
+}
+```
+
+While Mill cannot make the complexity of your build disappear, what it can do is give you
+the tools necessary to manage it. In this case, it means the JNI-related logic can be defined
+once and re-used, rather than duplicating it as is often done in Maven `pom.xml`s.
+
+== Extensibility
+
+Even though Maven is designed to be declarative, in many real-world codebases you end
+up needing to run ad-hoc scripts and logic. This section will explore two such scenarios,
+so you can see how Mill differs from Maven in the handling of these requirements.
+
+=== Groovy
+
+The Maven build for the `common/` subproject
+uses a Groovy script for code generation. This is configured via:
+
+```xml
+<properties>
+  <collection.template.dir>${project.basedir}/src/main/templates</collection.template.dir>
+  <collection.template.test.dir>${project.basedir}/src/test/templates</collection.template.test.dir>
+  <collection.src.dir>${project.build.directory}/generated-sources/collections/java</collection.src.dir>
+  <collection.testsrc.dir>${project.build.directory}/generated-test-sources/collections/java</collection.testsrc.dir>
+</properties>
+<plugin>
+  <groupId>org.codehaus.gmaven</groupId>
+  <artifactId>groovy-maven-plugin</artifactId>
+  <version>2.1.1</version>
+  <dependencies>
+    <dependency>
+      <groupId>org.codehaus.groovy</groupId>
+      <artifactId>groovy</artifactId>
+      <version>3.0.9</version>
+    </dependency>
+    <dependency>
+      <groupId>ant</groupId>
+      <artifactId>ant-optional</artifactId>
+      <version>1.5.3-1</version>
+    </dependency>
+  </dependencies>
+  <executions>
+    <execution>
+      <id>generate-collections</id>
+      <phase>generate-sources</phase>
+      <goals>
+        <goal>execute</goal>
+      </goals>
+      <configuration>
+        <source>${project.basedir}/src/main/script/codegen.groovy</source>
+      </configuration>
+    </execution>
+  </executions>
+</plugin>
+```
+
+In contrast, the Mill build configures the code generation as follows:
+
+```scala
+import $ivy.`org.codehaus.groovy:groovy:3.0.9`
+import $ivy.`org.codehaus.groovy:groovy-ant:3.0.9`
+import $ivy.`ant:ant-optional:1.5.3-1`
+
+object common extends NettyModule{
+  ...
+  def script = T.source(millSourcePath / "src" / "main" / "script")
+  def generatedSources0 = T{
+    val shell = new groovy.lang.GroovyShell()
+    val context = new java.util.HashMap[String, Object]
+
+    context.put("collection.template.dir", "common/src/main/templates")
+    context.put("collection.template.test.dir", "common/src/test/templates")
+    context.put("collection.src.dir", (T.dest / "src").toString)
+    context.put("collection.testsrc.dir", (T.dest / "testsrc").toString)
+
+    shell.setProperty("properties", context)
```

Comment:
I'd like to leave this as is, since we're interoping with Java libraries, and this example is meant to be readable to Java programmers. Scala is more idiomatic Mill, but staying in `java.util.HashMap`-land will be much more familiar and less scary

Reply:
I almost expected this answer as I had the same thought. ;)